src/xtgeo/grid3d/_roff_parameter.py

```diff
@@ -0,0 +1,264 @@
+        Args:
+            filelike (str or byte stream): The file to write to.
+            roff_format (roffio.Format): The format to write the file in.
+        """
+        data = OrderedDict(
+            {
+                "filedata": {"filetype": "parameter"},
+                "dimensions": {"nX": self.nx, "nY": self.ny, "nZ": self.nz},
+                "parameter": {"name": self.name},
+            }
+        )
+        if self.code_names is not None:
+            data["parameter"]["codeNames"] = list(self.code_names)
+        if self.code_values is not None:
+            data["parameter"]["codeValues"] = self.code_values
+        data["parameter"]["data"] = self.values
+        roffio.write(filelike, data)
+
+    @staticmethod
+    def from_file(filelike, name=None):
+        """
+        Read a RoffParameter from a roff file
+        Args:
+            filelike (str or byte stream): The file to read from.
+            name(str or None): The name of the parameter to get from the file,
+                if name=None, returns the first parameter read.
+        Returns:
+            The RoffGrid in the roff file.
+        """
+
+        def should_skip_parameter(tag, tagkey, keyvalue):
+            if tag == "parameter" and tagkey == "name":
+                if name is None or keyvalue == name:
+                    return False
+                return True
+            return False
+
+        translate_kws = {
+            "dimensions": {"nX": "nx", "nY": "ny", "nZ": "nz"},
+            "parameter": {
+                "name": "name",
+                "data": "values",
+                "codeValues": "code_values",
+                "codeNames": "code_names",
+            },
+        }
+        optional_keywords = defaultdict(
+            list,
+            {"parameter": ["codeValues", "codeNames"]},
+        )
+        found = {
+            tag_name: {key_name: None for key_name in tag_keys.keys()}
+            for tag_name, tag_keys in translate_kws.items()
+        }
+        found["filedata"] = {"filetype": None}
+        with roffio.lazy_read(filelike) as tag_generator:
+            for tag, keys in tag_generator:
+                if tag in found:
+                    if tag == "parameter" and found["parameter"]["data"] is not None:
+                        # We have already found the right parameter so skip
+                        # reading and potentially overwriting
+                        continue
+                    for key in keys:
+                        if should_skip_parameter(tag, key[0], key[1]):
+                            # Found a parameter, but not the one we are looking for
+                            # reset and look on
+                            for key_name in found["parameter"].keys():
+                                found["parameter"][key_name] = None
+                            break
+                        if key[0] in found[tag]:
+                            if found[tag][key[0]] is not None:
```

Comment:
The trick was to populate only the tags that we are interested in and skip the rest. So if `tag not in found` it is not a relevant `tag`. I added some explaining text for the dictionary.

Reply:
Yeh, I got what the code did, I just figured it would read better with slightly different naming. One could perhaps have made the `found` simply empty (default) dict, and instead made `valid_tags = translate_kws.keys()`, and checked `if tag in valid_tags`.. populate `found`. Then one wouldn't get the "how can `tag` be in `found` if it has to be already been "found" in order to populate it??"..

It's not very important! :P